packages/Popover/src/Popover.tsx

```diff
@@ -0,0 +1,626 @@
+}
+
+export default class Popover extends Component<Props, State> {
+  static displayName = 'Popover';
+
+  static propTypes = {
+    children: PropTypes.node,
+    active: PropTypes.bool,
+    className: PropTypes.string,
+    align: PropTypes.oneOf(['top', 'bottom', 'left', 'right']),
+    justify: PropTypes.oneOf(['start', 'middle', 'end']),
+    refEl: PropTypes.object,
+    withoutPortal: PropTypes.bool,
+    getUpdatePosition: PropTypes.func,
+  };
+
+  static defaultProps: DefaultProps = {
+    align: 'bottom',
+    justify: 'start',
+    active: false,
+  };
+
+  state = {
+    windowHeight: window.innerHeight,
+    windowWidth: window.innerWidth,
+    hasMounted: false,
+    referenceElPos: defaultRefPosition,
+    contentElPos: defaultRefPosition,
+    referenceElement: null,
+  };
+
+  componentDidMount() {
+    this.setReferenceElement();
+    this.setState({ hasMounted: true });
+
+    window.addEventListener('resize', this.handleWindowResize);
+  }
+
+  componentDidUpdate(prevProps: Props, prevState: State) {
+    const { align, justify, active } = this.props;
+    const {
+      windowWidth,
+      windowHeight,
+      referenceElement,
+      referenceElPos,
+    } = this.state;
+
+    const posPropsUpdated =
+      prevProps.active !== active ||
+      prevProps.align !== align ||
+      prevProps.justify !== justify;
+
+    const windowUpdated =
+      prevState.windowWidth !== windowWidth ||
+      prevState.windowHeight !== windowHeight;
+
+    if (!referenceElement || !referenceElPos) {
+      this.setReferenceElement();
+    }
+
+    if (posPropsUpdated || windowUpdated) {
+      this.updateReferencePositions();
+    }
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('resize', this.handleWindowResize);
+  }
+
+  handleWindowResize = () => {
+    this.setState(
+      {
+        windowHeight: window.innerHeight,
+        windowWidth: window.innerWidth,
+      },
+      () => {
+        this.setContentElPosition();
+      },
+    );
+  };
+
+  updateReferencePositions() {
+    const { referenceElement } = this.state;
+    const newReferenceElPos = this.getRefPosition(referenceElement);
+    const contentEl = this.contentRef && this.contentRef.current;
+
+    if (!contentEl) {
+      this.setState({
+        referenceElPos: newReferenceElPos,
+      });
+
+      return;
+    }
+
+    this.setState({
+      contentElPos: this.getRefPosition(contentEl),
+      referenceElPos: newReferenceElPos,
+    });
+  }
+
+  // Sets the element to position relative to based on passed props,
+  // and stores it, and it's position in state.
+  setReferenceElement() {
+    const { refEl: passedRef } = this.props;
+    const { referenceElement, referenceElPos } = this.state;
+
+    if (referenceElement) {
+      if (!referenceElPos) {
+        this.setState({
+          referenceElPos: this.getRefPosition(referenceElement),
+        });
+      }
+
+      return;
+    }
+
+    const newReferenceElement = (() => {
```

Comment:
Any need for this to be an IIFE?

Reply:
Originally, this was written as

```JS
  let newReferenceElement = null

  // Various conditionals and such to give it a value
```

So no, it doesn't strictly *need* to be an IIFE, but writing it as one lets us declare the variable as the result of a block expression. This is why I really want JS `do`.

Reply:
We reuse the variable, so making it strictly a function means we would have to call it twice in the function, which seems like a performance no-no since it is dealing with DOM nodes.